### ISSUE TITLE
Code-gen bug related to 32-bit floats (actually: rand())

### ISSUE THREAD

**Issue by floooh** [WebAssembly/binaryen#817]
TL;DR: while writing the bug descriptions below I noticed that ~~most of the time~~ _always_ random numbers are involved... always through glm::linearRand() or glm::ballRand(), which in turn call **std::rand()**.

I have rebuilt the Oryol samples with emscripten incoming from 31-Oct-2016 (without separate spidemonkify and binaryen builds, see build details below). I'm now seeing differences in some of the Oryol demos which seem to be related to 32-bit floating math. The asm.js (and PNaCl) versions of the demos look correct. Firefox and Chrome show the same results (haven't tested in Edge).

Here are links to the demos where I noticed differences:

Vertex colors are too dark: http://floooh.github.io/oryol/wasm/Shapes.html. The colors are written as 4 floating point components, and created with glm::linearRand(), code is here: https://github.com/floooh/oryol/blob/master/code/Modules/Assets/Gfx/ShapeBuilder.cc#L276

Vertex colors too dark, and particle emission directions limited to 1 quadrant: http://floooh.github.io/oryol/wasm/Instancing.html, vertex colors are again created with glm::linearRand(), emission directions with glm::ballRand() (here: https://github.com/floooh/oryol/blob/master/code/Samples/Instancing/Instancing.cc#L113).

Weird behaviour for falling characters: http://floooh.github.io/oryol/wasm/DebugText.html, for the initial position of characters again glm::linearRand() is involved: http://floooh.github.io/oryol/wasm/DebugText.html

Now it gets a bit weirder:

Again, too dark colors and initial physics velocity to small, but also specular highlight broken on some shapes: http://floooh.github.io/oryol-samples/wasm/BulletPhysicsBasic.html. For the initial velocity, glm::ballRand() is used, for the colors glm::linearRand(). The broken specular highlights are a bit weird, maybe this happens when one of the color channels becomes negative.

Broken colors again: http://floooh.github.io/oryol-samples/wasm/StbVoxelDemo.html, glm::linearRand().

And the last one: http://floooh.github.io/oryol-samples/wasm/BulletPhysicsCloth.html, this one only has broken specular highlights and wrong colors like the other physics demo.

**Comment by kripken** [comment]
Does this happen in an `-O0` build? If so it could be an optimizer bug, which is the most likely thing.

In any case, if this worked previously as wasm, then we should bisect. I can do that, if it's easy to build one of your testcases? Maybe you can set up a repo with just the bitcode, asset files, and a shell script that calls emcc on the bitcode?

**Comment by floooh** [comment]
Compiling with```-O0``` I'm getting a binaryen assertion in the linker stage:

```
Assertion failed: (double(ret) == getNumber()), function getInteger, file /Users/floh/.emscripten_ports/binaryen/binaryen-version_18/src/emscripten-optimizer/simple_ast.h, line 231.
```
This happens when compiling the demos that have the bug, but not in the others that work correctly. 

I'll try to build a minimal test case tomorrow to simplify the bisect.

Another thing that occured to me: std::rand() is integer, but the results are used as float. Maybe the problem is not rand() itself, but the casting from int to float?

The emscripten linker command line is:
```
/Users/floh/projects/fips-sdks/osx/emsdk_portable/emscripten/incoming/em++    --em-config /Users/floh/projects/fips-sdks/osx/emsdk_portable/.emscripten --cache /Users/floh/projects/fips-sdks/osx/emsdk_portable/.emscripten_cache  -fno-exceptions -fno-rtti -std=c++11 -stdlib=libc++ -fstrict-aliasing -Wall -Wno-multichar -Wextra -Wno-unused-parameter -Wno-unknown-pragmas -Wno-ignored-qualifiers -Wno-long-long -Wno-overloaded-virtual -Wno-deprecated-writable-strings -Wno-unused-volatile-lvalue -Wno-inconsistent-missing-override -Wno-warn-absolute-paths -Wno-expansion-to-defined -O0 -g -D_DEBUG_ -D_DEBUG -DFIPS_DEBUG=1   --em-config /Users/floh/projects/fips-sdks/osx/emsdk_portable/.emscripten --cache /Users/floh/projects/fips-sdks/osx/emsdk_portable/.emscripten_cache  --memory-init-file 1 -s TOTAL_MEMORY=134217728 -s ERROR_ON_UNDEFINED_SYMBOLS=1 -s EXPORTED_FUNCTIONS="['_main','_enter_fullscreen','_enter_soft_fullscreen']" -s NO_EXIT_RUNTIME=1 -s NO_FILESYSTEM=1 -s DISABLE_EXCEPTION_CATCHING=1 -s WASM=1 -s 'BINARYEN_METHOD="native-wasm"'  -O0 -g CMakeFiles/Instancing.dir/Instancing.cc.o CMakeFiles/Instancing.dir/shaders.cc.o  -o /Users/floh/projects/fips-deploy/oryol/wasm-make-debug/Instancing.html  ../../Modules/Gfx/libGfx.a ../../Modules/Assets/libAssets.a ../../Modules/Dbg/libDbg.a ../../Modules/Input/libInput.a ../../Modules/Gfx/libGfx.a ../../Modules/Resource/libResource.a ../../Modules/Core/libCore.a
```

**Comment by floooh** [comment]
Ok, so here's a minimal repro case. I had to use one of the Oryol samples, the problem doesn't seem to happen in a simple hello-world-style test.

To reproduce the problem:
```
> git clone https://github.com/floooh/wasm_test_dump
> cd wasm_test_dump
> ./minrepro.sh
...
Assertion failed: (double(ret) == getNumber()), function getInteger, file /Users/floh/.emscripten_ports/binaryen/binaryen-version_18/src/emscripten-optimizer/simple_ast.h, line 231.
...
```

Ignore all the 'unresolved symbol' warnings, I only added the minimal set of .bc libs which are required to generate that assertion. 

One interesting thing: if the order of Shapes.bc and libAssets.bc is reversed in the compile command line, the assert doesn't happen.

So, running this triggers the assert:
```
em++ Shapes.bc libAssets.bc -o Shapes.html -s WASM=1 -O0
```
But running this doesn't:
```
em++ libAssets.bc Shapes.bc -o Shapes.html -s WASM=1 -O0
```
Do you have time to do a bisect? If not, what would be the recommended procedure to set up an automatic bisect for binaryen, do you have a script for this somewhere?

I might not have time until next week to do this though :/

**Comment by floooh** [comment]
Another interesting thing I just noticed: leaving away the -O0, or compiling with -O3 also triggers the assert with the .bc files in the repo... this doesn't happen though when the entire build is compiled with -O3 (so that the bitcode object files are already compiled with -O3).

I'm assuming/hoping that the assert and the code-gen-problem with optimization on have the same cause (so that if the assert is no longer triggered, the optimized version would also be correct). If the 2 problems are unrelated I probably need to add a few more things to the repo (such as .bc files with different optimization levels).

**Comment by kripken** [comment]
Ok, #820 fixes that assert. It's possible there is a second problem though, I have no strong reason to assume this was the cause of the other issue you saw.

**Comment by floooh** [comment]
Ok, I'll give it a try and report back. Is the fix already in the incoming emscripten SDK or do I need to setup a separate binaryen like in the 'old days'?

**Comment by kripken** [comment]
Like the old days, sorry - it's in binaryen master, but I don't update the emscripten version of binaryen before I run all the tests, so it's not there yet.

**Comment by floooh** [comment]
You're assumption was right, the assertion is gone, but the original problem's is still there, at least it's confirmed that it's also happening with ```-O0``` builds, not just ```-O3```. I'll update the test-case git repo with new files so that a complete .html can be built and tested. I'll also try to find out what's wrong with those float numbers (they should be in the range 0.0 to 1.0 for the vertex colors, but seem to be smaller by at least half, or even go into negative range).

**Comment by floooh** [comment]
Ok, git repo is updated (the bitcode files are actually exactly the same).

There's 2 build scripts, one to build the WASM version, and one for the asm.js version:

```bash
# build the wasm version:
> ./build_wasm.sh
# build the asmjs version
> ./build_asmjs.sh 
```

This builds the files **Shapes_wasm.html** and **Shapes_asmjs.html**.

The faulty WASM version should look like this (vertex colors too dark):

<img width="670" alt="screen shot 2016-11-02 at 19 35 38" src="https://cloud.githubusercontent.com/assets/1699414/19942307/98433f96-a133-11e6-8698-b48727efc5ff.png">

And the correct asm.js version like this:

<img width="654" alt="screen shot 2016-11-02 at 19 36 37" src="https://cloud.githubusercontent.com/assets/1699414/19942344/b5ddccce-a133-11e6-9bff-2cf38055bad3.png">

I'll try to do a bisect on binaryen (now that I have binaryen running from the repo again it should be trivial), but I'm traveling and might only be able to go back to this in a couple of days.

**Comment by floooh** [comment]
Ok, here's some interesting new info: all vertex colors above 0.5 are instead negative. It looks like a signed/unsigned mixup of the original integer random numbers.

For instance, here are the first 5 vertices for the cube in the asm.js demo (those values are the correct ones):

```
x: 0.000547, y: 0.088207, z: 0.404341, w: 1.000000
x: 0.124982, y: 0.585633, z: 0.471635, w: 1.000000
x: 0.697425, y: 0.383120, z: 0.843062, w: 1.000000
x: 0.112603, y: 0.830233, z: 0.155412, w: 1.000000
x: 0.776852, y: 0.631963, z: 0.740406, w: 1.000000
```

And here the same vertices in the broken WASM demo, note how all values 'v' above 0.5 are instead (-1.0 + v)

```
x: 0.000547, y: 0.088207, z: 0.404341, w: 1.000000
x: 0.124982, y: -0.414367, z: 0.471635, w: 1.000000
x: -0.302575, y: 0.383120, z: -0.156938, w: 1.000000
x: 0.112603, y: -0.169767, z: 0.155412, w: 1.000000
x: -0.223148, y: -0.368037, z: -0.259594, w: 1.000000
```

The computation for those random numbers in GLM looks tricky (here: https://github.com/g-truc/glm/blob/2336264f4eecea8a029b14e54b275e77318c3782/glm/gtc/random.inl#L228):

```cpp
	template <template <class, precision> class vecType>
	struct compute_linearRand<float, highp, vecType>
	{
		GLM_FUNC_QUALIFIER static vecType<float, highp> call(vecType<float, highp> const & Min, vecType<float, highp> const & Max)
		{
			return vecType<float, highp>(compute_rand<uint32, highp, vecType>::call()) / static_cast<float>(std::numeric_limits<uint32>::max()) * (Max - Min) + Min;
		}
	};
```

And that compute_rand for uint32 also looks a bit non-trivial (it's building the 32 bit number from 2 16-bit numbers, here: https://github.com/g-truc/glm/blob/2336264f4eecea8a029b14e54b275e77318c3782/glm/gtc/random.inl#L105

```cpp
	template <precision P, template <class, precision> class vecType>
	struct compute_rand<uint32, P, vecType>
	{
		GLM_FUNC_QUALIFIER static vecType<uint32, P> call()
		{
			return
				(vecType<uint32, P>(compute_rand<uint16, P, vecType>::call()) << static_cast<uint32>(16)) |
				(vecType<uint32, P>(compute_rand<uint16, P, vecType>::call()) << static_cast<uint32>(0));
		}
	};
```

... and the 16-bit number in turn is built by or-ing 2 8-bit numbers... and finally an 8-bit random number is built like this:

```cpp
std::rand()) % std::numeric_limits<uint8>::max();
```

...maybe I can build a simplified test case with this in mind...

**Comment by kripken** [comment]
Interesting, thanks. Bisecting this on the repos, it's on
````
cdd52f71b2d7e24a7654d46d537c2260a73eaf5c is the first bad commit
commit cdd52f71b2d7e24a7654d46d537c2260a73eaf5c
Author: Alon Zakai <alonzakai@gmail.com>
Date:   Tue Sep 13 13:55:57 2016 -0700

    default PRECISE_F32 to on when compiling to wasm
````
Which is interesting, since enabling `PRECISE_F32` on asm.js does not change anything...

And in fact disabling `PRECISE_F32` for wasm does make things work. So we are doing something wrong with floats, indeed.

Very strange. I can investigate more, perhaps bisect on the codebase somehow. But maybe this helps your hunch, and you can craft a small testcase?

**Comment by kripken** [comment]
Turns out direct bisecting on two builds is hard, as `PRECISE_F32` changes the ABI.

Maybe we can bisect on the history of oryol - if we go back, does it start working? Or is this an old demo that has not changed in very long?

Alternatively, we can use shell logging here, if it's feasible to make a shell version of this testcase?

**Comment by kripken** [comment]
Ok, I got autodebug logging working in the browser, and diffed builds with and without `PRECISE_F32`. The log shows the first differences happens in

````
_ZN3glm5tvec3IfLNS_9precisionE0EEC2IjLS1_0EEERKNS0_IT_XT0_EEE
glm::tvec3<float, (glm::precision)0>::tvec3<unsigned int, (glm::precision)0>(glm::tvec3<unsigned int, (glm::precision)0> const&)
````

and indeed it involves f32 operations.

Now, looking at this fairly short method, the wasts differ only in one using f32s and the other using f64s, and then there are the necessary promotions/demotions to f32 for storing (if we use f64s), etc. So I don't see any bad code being generated. Is it possible the values here are just in the range where f32/f64 differences matter, and the method is sensitive to them? Looks like it loads an i32, converts to f32 or f64, then does the same for another i32 4 bytes after it. It's the second of them where the difference occurs,

````
-AD:49319,378846291
+AD:49319,378846304
````

49319 is the id, the value after is the f32 or f64 value. So the difference between the two is indeed fairly small, and could be a rounding difference in f32/f64. Maybe those accumulate?

**Comment by kripken** [comment]
Oh, I was diffing the wrong thing. Diffing f32 vs f64 shows rounding differences that are ok. But diffing asm.js with f32 to wasm with f32 shows the problem: we didn't do fround of unsigned properly. This happens in that same method mentioned above actually. Fix is in #821.

Building with that, looks like the program renders correctly.

Thanks @floooh!

**Comment by binji** [comment]
Probably worth adding a spec test that would catch this bug too.

**Comment by kripken** [comment]
Oh, this was specific to asm2wasm, I don't think a spec test could have helped. The spec suite already has tests for both signed and unsigned i32-to-f32 conversions.

**Comment by kripken** [comment]
Actually if we had a wasm to asm.js convertor, then those tests would have found this, by checking round-trips, hmm.

**Comment by floooh** [comment]
All working fine now, I have updated the WebAssembly samples at http://floooh.github.io/oryol/ and http://floooh.github.io/oryol-samples/.

Thanks!

**Comment by floooh** [comment]
PS: Any ETA when the fixed Binaryen will be in the emscripten incoming SDK?

**Comment by kripken** [comment]
It's on incoming now, so getting latest incoming from the SDK should be fine.
